src/ripple/app/tx/impl/DepositPreauth.cpp

```diff
@@ -0,0 +1,195 @@
+        return temMALFORMED;
+    }
+
+    // Make sure that the passed account is valid.
+    AccountID const target {auth.is_initialized() ? *auth : *unauth};
+    if (target == beast::zero)
+    {
+        JLOG(j.trace()) <<
+            "Malformed transaction: Authorized or Unauthorized field zeroed.";
+        return temINVALID_ACCOUNT_ID;
+    }
+
+    // An account may not preauthorize itself.
+    if (auth.is_initialized() && (target == ctx.tx[sfAccount]))
+    {
+        JLOG(j.trace()) <<
+            "Malformed transaction: Attempting to DepositPreauth self.";
+        return temCANNOT_PREAUTH_SELF;
+    }
+
+    return preflight2 (ctx);
+}
+
+TER
+DepositPreauth::preclaim(PreclaimContext const& ctx)
+{
+    // Determine which operation we're performing: authorizing or unauthorizing.
+    if (ctx.tx.isFieldPresent (sfAuthorize))
+    {
+        // Verify that the Authorize account is present in the ledger.
+        AccountID const auth {ctx.tx[sfAuthorize]};
+        auto const sleAuth = ctx.view.read (keylet::account (auth));
+        if (!sleAuth)
+            return tecNO_TARGET;
+
+        // Verify that the Preauth entry they asked to add is not already
+        // in the ledger.
+        if (ctx.view.exists (
+            keylet::depositPreauth (ctx.tx[sfAccount], auth)))
+            return tecDUPLICATE;
+    }
+    else
+    {
+        // Verify that the Preauth entry they asked to remove is in the ledger.
+        AccountID const unauth {ctx.tx[sfUnauthorize]};
+        if (! ctx.view.exists (
+            keylet::depositPreauth (ctx.tx[sfAccount], unauth)))
+            return tecNO_ENTRY;
+    }
+    return tesSUCCESS;
+}
+
+TER
+DepositPreauth::doApply ()
+{
+    auto const sleOwner = view().peek (keylet::account (account_));
+
+    // Determine which operation we're performing: authorizing or unauthorizing.
+    if (ctx_.tx.isFieldPresent (sfAuthorize))
+    {
+        // A preauth counts against the reserve of the issuing account, but we
+        // check the starting balance because we want to allow dipping into the
+        // reserve to pay fees.
+        {
+            STAmount const reserve {view().fees().accountReserve (
+                sleOwner->getFieldU32 (sfOwnerCount) + 1)};
+
+            if (mPriorBalance < reserve)
+                return tecINSUFFICIENT_RESERVE;
+        }
+
+        // Preclaim already verified that the Preauth entry does not yet exist.
+        // Create and populate the Preauth entry.
+        AccountID const auth {ctx_.tx[sfAuthorize]};
+        uint256 const preauthIndex {getDepositPreauthIndex (account_, auth)};
+        auto slePreauth =
+            std::make_shared<SLE>(ltDEPOSIT_PREAUTH, preauthIndex);
+
+        slePreauth->setAccountID (sfAccount, account_);
+        slePreauth->setAccountID (sfAuthorize, auth);
+        view().insert (slePreauth);
+
+        auto viewJ = ctx_.app.journal ("View");
+        auto const page = dirAdd (view(), keylet::ownerDir (account_),
```

Comment:
For new code, prefer `auto const page = view().dirInsert (keylet::ownerDir (account_), preauth->key(), describeOwnerDir (account_));`